embeddings/embeddings.py

```diff
@@ -0,0 +1,242 @@
+        if precision_nbits == 32:
+            print("Keeping in fp32 precision")
+        else:
+            print(f"Quantizing to {precision_nbits} bits")
+            nn.QuantizedLinear.quantize_module(mlx_model, bits=precision_nbits)
+        tokenizer = AutoTokenizer.from_pretrained(model_path)
+
+        return mlx_model, tokenizer
+
+
+class EmbeddingModel:
+    def __init__(
+        self,
+        model_path: str, # path to model on huggingface. must be BERT
+        max_length: int = 512,
+        pooling_strategy: Literal["mean", "cls"] = "cls",
+        normalize: bool = True,
+        precision_nbits: int = 8
+    ):
+        super().__init__()
+        self.model, self.tokenizer = Bert.from_pretrained(model_path, precision_nbits)
+        self.max_length = max_length
+        self.pooling_strategy = pooling_strategy
+        self.normalize = normalize
+    
+    def encode(self, sentences, batch_size=32, show_progress=True, **kwargs):
+        """
+        Returns a list of embeddings for the given sentences.
+        Args:
+            sentences (`List[str]`): List of sentences to encode
+            batch_size (`int`): Batch size for the encoding
+
+        Returns:
+            `List[np.ndarray]` or `List[tensor]`: List of embeddings for the given sentences
+        """
+        result = None
+        if show_progress:
+            pbar = tqdm.tqdm(total=len(sentences))
+        else:
+            pbar = None
+        for i in range(0, len(sentences), batch_size):
+            batch = sentences[i : i + batch_size]
+            tokenized = self.tokenizer(
+                batch,
+                padding="longest",
+                truncation=True,
+                max_length=self.max_length,
+                return_tensors="np",
+            )
+            hidden_states, pooler_output = self.model(
+                input_ids=mx.array(tokenized["input_ids"]),
+                token_type_ids=mx.array(tokenized["token_type_ids"]) if "token_type_ids" in tokenized else None,
+                attention_mask=mx.array(tokenized["attention_mask"]),
+            )
+            hidden_states, pooler_output = np.array(hidden_states), np.array(pooler_output)
+            attn_mask = np.array(tokenized["attention_mask"])
+            if self.pooling_strategy == "mean":
+                pooled = np.sum(hidden_states * np.expand_dims(attn_mask, -1), axis=1) / np.sum(attn_mask, axis=1, keepdims=True)
+            else:
+                pooled = pooler_output
+            if self.normalize:
+                pooled = pooled / np.linalg.norm(pooled, axis=1, keepdims=True)
+
+            if result is None:
+                result = pooled
+            else:
+                result = np.concatenate([result, pooled], axis=0)
+            if pbar is not None:
+                pbar.update(len(batch))
+
+        return result
+
+def test_embeddings():
+    mx_model, tokenizer = Bert.from_hugging_face("BAAI/bge-small-en-v1.5", precision_nbits=16)
```

Comment:
Is this implemented

Reply:
Good catch on the renamed thing! I might need help to get the 16-bits version working. I found that when I naively cast all the weights of the model to 16 bits upon loading it, performance plummeted, which is why I only made 32 bit or quantized an option (the quantization utils didn't support 16 bit "quantization" because normally you don't need special scales etc. for 16 bits).

Reply:
When using 32 bits there's still a small difference between sentence-transformers and the MLX version, ~2% in cosine similarity. Is this expected?

Reply:
I'm not sure why that is. All I verify in my testing is that MTEB performance is preserved on a subset of tasks.